File: src/Board.py
from src.Colour import Colour
from src.Tile import Tile


class Board:
    """Class that describes the Hex board."""

    _size: int
    _tiles: list[list[Tile]]
    _winner: Colour | None

    def __init__(self, board_size=11):
        self._size = board_size

        self._tiles = []
        for i in range(board_size):
            new_line = []
            for j in range(board_size):
                new_line.append(Tile(i, j))
            self._tiles.append(new_line)

        self._winner = None

    def __str__(self) -> str:
        return self.print_board()

    def __eq__(self, value: object) -> bool:
        if not isinstance(value, Board):
            return False

        if self._size != value.size:
            return False

        for i in range(self._size):
            for j in range(self._size):
                if self.tiles[i][j].colour != value.tiles[i][j].colour:
                    return False

        return True

    def from_string(string_input, board_size=11):
        """Loads a board from a string representation. If bnf=True, it will
        load a protocol-formatted string. Otherwise, it will load from a
        human-readable-formatted board.
        """

        b = Board(board_size=board_size)

        lines = [line.strip() for line in string_input.split("\n")]
        for i, line in enumerate(lines):
            chars = line.split(" ")
            for j, char in enumerate(chars):
                b.tiles[i][j].colour = Colour.from_char(char)
        return b

    def has_ended(self, colour: Colour = None):
        """Checks if the game has ended. It will attempt to find a red chain
        from top to bottom or a blue chain from left to right of the board.
        """

        # Red
        # for all top tiles, check if they connect to bottom
        if colour == Colour.RED:
            for idx in range(self._size):
                tile = self._tiles[0][idx]
                if not tile.is_visited() and tile.colour == Colour.RED and self._winner is None:
                    self.DFS_colour(0, idx, Colour.RED)
        # Blue
        # for all left tiles, check if they connect to right
        elif colour == Colour.BLUE:
            for idx in range(self._size):
                tile = self._tiles[idx][0]
                if not tile.is_visited() and tile.colour == Colour.BLUE and self._winner is None:
                    self.DFS_colour(idx, 0, Colour.BLUE)
        else:
            raise ValueError("Invalid colour")

        # un-visit tiles
        self.clear_tiles()

        return self._winner is not None

    def clear_tiles(self):
        """Clears the visited status from all tiles."""

        for line in self._tiles:
            for tile in line:
                tile.clear_visit()

    def DFS_colour(self, x, y, colour):
        """A recursive DFS method that iterates through connected same-colour
        tiles until it finds a bottom tile (Red) or a right tile (Blue).
        """

        self._tiles[x][y].visit()

        # win conditions
        if colour == Colour.RED:
            if x == self._size - 1:
                self._winner = colour
        elif colour == Colour.BLUE:
            if y == self._size - 1:
                self._winner = colour
        else:
            return

        # end condition
        if self._winner is not None:
            return

        # visit neighbours
        for idx in range(Tile.NEIGHBOUR_COUNT):
            x_n = x + Tile.I_DISPLACEMENTS[idx]
            y_n = y + Tile.J_DISPLACEMENTS[idx]
            if x_n >= 0 and x_n < self._size and y_n >= 0 and y_n < self._size:
                neighbour = self._tiles[x_n][y_n]
                if not neighbour.is_visited() and neighbour.colour == colour:
                    self.DFS_colour(x_n, y_n, colour)

    def print_board(self) -> str:
        """Returns the string representation of a board with color-coded text."""
        
        output = ""
        leading_spaces = ""
        for line in self._tiles:
            output += leading_spaces
            leading_spaces += " "
            for tile in line:
                # Apply color based on player
                colour = Colour.get_char(tile.colour)
                if colour == "R":
                    output += "\033[91mR\033[0m "  # Red for player1
                elif colour == "B":
                    output += "\033[94mB\033[0m "  # Blue for player2
                else:
                    output += "X "  # Default color for other tiles
            output += "\n"

        return output

    def get_winner(self) -> Colour:
        return self._winner

    @property
    def size(self) -> int:
        return self._size

    @property
    def tiles(self) -> list[list[Tile]]:
        return self._tiles

    def set_tile_colour(self, x, y, colour) -> None:
        self.tiles[x][y].colour = colour


if __name__ == "__main__":
    b = Board.from_string(
        "0R000B00000,0R000000000,0RBB0000000,0R000000000,0R00B000000,"
        + "0R000BB0000,0R0000B0000,0R00000B000,0R000000B00,0R0000000B0,"
        + "0R00000000B",
        bnf=True,
    )
    b.print_board(bnf=False)
    print(b.has_ended(), b.get_winner())

File: agents/SamAgent/Board.cpp
#include "Board.h"
#include <iostream>
#include <sstream>

// Constructor
Board::Board(int boardSize) : _size(boardSize), _winner(static_cast<Colour>(-1)) {
    for (int i = 0; i < boardSize; ++i) {
        std::vector<Tile> row;
        for (int j = 0; j < boardSize; ++j) {
            row.emplace_back(Tile(i, j));
        }
        _tiles.emplace_back(row);
    }
}

Board Board::clone() const {
    Board newBoard(_size); 
    newBoard._winner = _winner; 

    // Deep copy of tiles
    for (int i = 0; i < _size; ++i) {
        for (int j = 0; j < _size; ++j) {
            newBoard._tiles[i][j] = _tiles[i][j]; 
        }
    }

    return newBoard;
}

// Create a board from a string representation
Board Board::fromString(const std::string& stringInput, int boardSize) {
    Board board(boardSize);
    std::istringstream stream(stringInput);
    std::string line;
    int row = 0;

    while (std::getline(stream, line)) {
        int col = 0;
        for (char c : line) {
            if (c != ' ' && col < boardSize) {
                board._tiles[row][col].setColour(ColourUtil::fromChar(c));
                ++col;
            }
        }
        ++row;
    }
    return board;
}

// Perform depth-first search for connectivity
void Board::DFSColour(int x, int y, Colour colour) {
    _tiles[x][y].visit();

    // Win condition for RED (top-to-bottom) or BLUE (left-to-right)
    if ((colour == Colour::RED && x == _size - 1) ||
        (colour == Colour::BLUE && y == _size - 1)) {
        _winner = colour;
        return;
    }

    // Visit neighbours
    for (int i = 0; i < Tile::NEIGHBOUR_COUNT; ++i) {
        int nx = x + Tile::I_DISPLACEMENTS[i];
        int ny = y + Tile::J_DISPLACEMENTS[i];

        if (nx >= 0 && nx < _size && ny >= 0 && ny < _size) {
            Tile& neighbour = _tiles[nx][ny];
            if (!neighbour.isVisited() && neighbour.getColour() == colour) {
                DFSColour(nx, ny, colour);
                if (_winner != static_cast<Colour>(-1)) return;  // Early exit if a winner is found
            }
        }
    }
}

// Clears the visited status from all tiles
void Board::clearTiles() {
    for (auto& row : _tiles) {
        for (auto& tile : row) {
            tile.clearVisit();
        }
    }
}

// Check if the game has ended
bool Board::hasEnded(Colour colour) {
    if (colour == Colour::RED) {
        // Check top-to-bottom connection for RED
        for (int col = 0; col < _size; ++col) {
            if (!_tiles[0][col].isVisited() && _tiles[0][col].getColour() == Colour::RED && _winner == static_cast<Colour>(-1)) {
                DFSColour(0, col, Colour::RED);
            }
        }
    } else if (colour == Colour::BLUE) {
        // Check left-to-right connection for BLUE
        for (int row = 0; row < _size; ++row) {
            if (!_tiles[row][0].isVisited() && _tiles[row][0].getColour() == Colour::BLUE && _winner == static_cast<Colour>(-1)) {
                DFSColour(row, 0, Colour::BLUE);
            }
        }
    } else {
        throw std::invalid_argument("Invalid colour");
    }

    // Clear visited tiles
    clearTiles();

    return _winner != static_cast<Colour>(-1);
}

// Prints the board
std::string Board::printBoard() const {
    std::ostringstream output;
    std::string leadingSpaces = "";

    for (const auto& row : _tiles) {
        output << leadingSpaces;
        leadingSpaces += " ";
        for (const auto& tile : row) {
            char c = ColourUtil::getChar(tile.getColour());
            if (c == 'R') {
                output << "\033[91mR\033[0m ";  // Red
            } else if (c == 'B') {
                output << "\033[94mB\033[0m ";  // Blue
            } else {
                output << "X ";  // Default
            }
        }
        output << "\n";
    }

    return output.str();
}

// Getters
int Board::getSize() const {
    return _size;
}

const std::vector<std::vector<Tile>>& Board::getTiles() const {
    return _tiles;
}

Colour Board::getWinner() const {
    return _winner;
}

// Sets the colour of a tile
void Board::setTileColour(int x, int y, Colour colour) {
    _tiles[x][y].setColour(colour);
}
